### ISSUE TITLE
Stop gobbling exceptions in the serialization

### ISSUE THREAD

**Issue by avsm** [mirage/ocaml-dns#37]
Related to #36; in lib/protocol.ml we have:

```
let contain_exc l v =
  try
    Some (v ())
  with exn ->
    Printexc.print_backtrace stderr;
    Printf.eprintf "dns %s exn: %s\n%!" l (Printexc.to_string exn);
    None
```

This prints the exception to stderr (which is bad, see #36), and then returns a blank option with no hint about the error.  It does make more sense to raise the exception here or use an `Or_error.t` style polymorphic return.  This would change the `Protocol.Server` types though, so I'm deferring this for the moment while fixing some of the easier ones in #36.

**Comment by hannesm** [comment]
(jumping in here) I just recently [started to use](https://github.com/hannesm/ns.nqsb.io) ocaml-dns, and agree that the interface could be improved: exceptions can be thrown in parse, the Loader interface is very imperative (not sure whether this can be avoided, or if the Trie always have to be imperative).

Also, the fine Logs library could be used to emit log messages (instead of printing to standard output or standard error).

In addition, the test library implements a whole network stack, and an in-memory KV_RO interface.  Both might be useful, but deserve to be in separate libraries.

**Comment by avsm** [comment]
fixed in #159 by rewriting the whole darn interface thanks to @hannesm
